### ISSUE TITLE
importComponent not working in Jest

### ISSUE THREAD

**Issue by herschel666** [xing/hops#772]
## This is my intent

- [ ] I want to report a bug
- [x] I want to request a feature or change
- [ ] I want to provide or change a feature

## The problem

As a good citizen of the (React) developer community I not only want to test my code thoroughly, but also make my component tests in the most integrative way possible. Thus I test as high as possible in the component tree. This leads the Jest runner to come across asynchronously loaded components, which again makes it break.

## Proposed solution

Add [`@untool/react`'s Babel plugin](https://github.com/untool/untool/blob/cb08a3b8406ce03623c0bd5368e54c180c8d6707/packages/react/lib/babel.js) to `jest-preset-hops`.

**Comment by toomuchdesign** [comment]
The issue is produced by [`importComponent` wrapper component](https://github.com/untool/untool/blob/6ce662f527d8d05d184676dc398372799182ca09/packages/react/lib/runtime.js#L40) being rendered in tests.

The first error I get comes from `react-router` complaining about `importComponent` (which is a router-connected component) being used outside a `Router` component.

Even if we found a workaround to this `react-router error we might still step into any further issue, therefore we need a more radical solution to handle `importComponent` wrapper component in tests.

## Possible solutions

### Intro
From `hops` point of view, we should decide whether **`importComponent` wrapper component** should be **rendered in the tests** tree **or skipped** to directly render the actual imported component.

### Jest mock
The actual `importComponent` implementation might be replaced with a **Jest mock** which **wraps and render the imported component**.

The mock implementation might be simply described in the **documentation** or **supplied** by `jest-preset-hops`.

### Babel
Another solution might consist of adding a custom `babel` transpilation step (test environment only) to replace every `importComponent` instance with a plain instance of the imported component.

### Any further idea?
...

**Comment by robin-drexler** [comment]
I think we should use the babel transformer for tsx files, even if we decide to mock `importComponent`.

Since we also babel [ts  files with hops](https://github.com/xing/hops/pull/769), the same should happen when in the test environment as otherwise you'd end up with syntax errors in tests that you do not encounter in your actual application.

**Comment by herschel666** [comment]
The discussion shifted over to the [(un)tooling layer](https://github.com/untool/untool/issues/268).

**Comment by herschel666** [comment]
Aaaaaaand … it's back! The minimal approach should do it for now. Later on we'll improve if the need arises.

**Comment by robin-drexler** [comment]
Btw. I think in Hops 12 / untool 2, we should remove the option to pass a string to `importComponent` and have consumers pass the arrow function import statement themselves.

`importComponent(() => import ("./foo”))`

This would have two advantages:

- no mock needed if we also ensure `importComponent` checks for the existence of global object before accessing them.

- editors would be able to resolve the imported component. In its current form, for editors it's just a string that you can't follow.

This could already be done right now in a non breaking manner (as @dmbch brought up), but without the removal of the string version, the mock would still be needed.

**Comment by herschel666** [comment]
>This could already be done right now in a non breaking manner (as @dmbch brought up), but without the removal of the string version, the mock would still be needed.

Maybe we should then deprecate the option to pass in a string with the next minor release and output a warning in dev env.

**Comment by dmbch** [comment]
With regards to that arrow function and editor support..: this is all about being able to click the path string, right? I do not really see how type/prop hints for the returned wrapper component could be inferred, because babel’ing would of course still be required (but then I am no TS expert). Does `react-loadable` or any of its alternatives manage to provide full-on TS-support for imported components?

Oh, and would we not only be able to capture the loading state during testing deterministically without another, potentially more elaborate, mock (no Jest expert neither)? After all, that arrow function would only be called in `componentDidMount`.

I do not wish to stand in the way of progress or convenience, but I would like to completely understand the reasoning behind the proposed changes...

**Comment by herschel666** [comment]
> I do not really see how type/prop hints for the returned wrapper component could be inferred, because babel’ing would of course still be required (but then I am no TS expert).

If I get this right, it will type-wise make no difference whether `importComponent` takes a string or a function. To get the types of an asynchronously loaded component one has to use Typescript's `import()`-function.

```typescript
const Text: React.SFC<ImportComponentProps> = importComponent('./text', 'Text');

const render = ({
  Component: import('./text').Text,
  ...props: any,
}): React.ReactType => {
  return <Component {...props} />;
};
```

@ZauberNerd Can you confirm?

**Comment by ZauberNerd** [comment]
@herschel666 @dmbch  here is an example that TypeScript can infer:
```typescript
const Foo = Import(async () => (await import('./foo')).Foo);
```
The problem is, that we are mixing compile- and runtime and that won't work.
Also: The type of `import()` is a built-in, because it does IO, so we __need__ to have that function actually in code. And also the export name can't be a string, because that would be runtime and TypeScript does not exist at runtime ;)
So the only way this would work is how I've shown in the example above, because that would be statically analyzable at compile time.

**Comment by herschel666** [comment]
Resolved by #798
